nixos/modules/programs/dconf.nix

```diff
@@ -1,55 +1,174 @@
+      nativeBuildInputs = [ pkgs.dconf ];
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
+  recuriseLock = with lib; settings: locks: (mapAttrsRecursive
+    (path: value: optionalAttrs value
+      (mapAttrsRecursiveCond (value: !lib.gvariant.isGVariant value)
+        (path: value: true)
+        (foldl'
+          (root: name: if hasAttr "${name}" root then root."${name}" else
+          throw "${concatStringsSep "." (tail path)} doesn't exist in settings.")
+          { v = settings; }
+          path)))
+    { v = locks; }).v;
+
+  # Generate Dconf DB from attrset and keyfiles
+  mkDconfDb = val: compileDconfDb (pkgs.symlinkJoin {
+    name = "nixos-generated-dconf-keyfiles";
+    paths = [
+      (pkgs.writeTextDir "nixos-generated-dconf-keyfiles" (lib.generators.toDconfINI val.settings))
+      (pkgs.writeTextDir "locks/nixos-generated-dconf-locks" (lib.generators.toDconfLocks
+        (lib.recursiveUpdate (recuriseLock val.settings val.lockSubpaths) val.locks)))
+      val.keyfiles
+    ];
+  });
+
+  # Generate Dconf profile
+  mkDconfProfile = name: value: pkgs.writeTextDir "etc/dconf/profile/${name}" (
+    if lib.isDerivation value || lib.isPath value then
+      lib.readFile value
+    else
+      lib.concatMapStrings (x: "${x}\n") ((
+        lib.optional value.enableUserDb "user-db:user"
+      ) ++ (
+        map
+          (value:
+            let
+              db = if lib.isAttrs value && !lib.isDerivation value then mkDconfDb value else value;
+            in
+            "file-db:${db}")
+          value.databases
+      ))
+  );
 in
 {
-  ###### interface
-
   options = {
     programs.dconf = {
-      enable = mkEnableOption (lib.mdDoc "dconf");
+      enable = lib.mkEnableOption (lib.mdDoc "dconf");
 
-      profiles = mkOption {
-        type = types.attrsOf types.path;
-        default = {};
-        description = lib.mdDoc "Set of dconf profile files, installed at {file}`/etc/dconf/profiles/«name»`.";
-        internal = true;
+      profiles = lib.mkOption {
+        type = with lib.types; attrsOf (oneOf [
+          path
+          package
+          (submodule {
+            options = {
+              enableUserDb = lib.mkOption {
+                type = bool;
+                default = true;
+                description = lib.mdDoc "Add `user-db:user` at the beginning of the profile.";
+              };
+
+              databases = lib.mkOption {
+                type = with lib.types; listOf (oneOf [
+                  path
+                  package
+                  (submodule {
+                    options = {
+                      keyfiles = lib.mkOption {
+                        type = with lib.types; listOf (oneOf [ path package ]);
+                        default = [ ];
+                        description = lib.mdDoc "A list of dconf keyfiles.";
+                      };
+                      settings = lib.mkOption {
+                        type = lib.types.attrs;
```

Comment:
Why `attrs` instead of just `attrsOf gvariant.value` like [home-manager has](https://nix-community.github.io/home-manager/options.html#opt-dconf.settings)? It would also be close to what comes from `dconf dump`.

Reply:
HM uses the path as the key but in this module the key is get from recursive attrs instead.

Reply:
Right. I am asking why was this decision made. Is it just to make locks implementation cleaner? Or are there some benefits for users?

Reply:
This design is inherited from https://github.com/NixOS/nixpkgs/pull/189099 and the locks implementation is based on this. I feel that this design can help organize the settings in a Nix style and the user can still choose to use the HM style.

> Is it just to make locks implementation cleaner?

Yes, looks like it's easier to implement the lockSubpaths in this design.

Reply:
Actually, one downside of nested attributes is that it no longer allows having both “files” and “directories“ of the same name, which `dconf` apparently supports. Not sure if any real application does this but if we want to allow creating arbitrary database, we probably cannot use this format.

Reply:
I'm not sure if I understand you correctly. Here is an example:

```
$ dconf dump /org/gnome/settings-daemon/plugins/
...
[media-keys]
...
custom-keybindings=['/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/']
...

[media-keys/custom-keybindings/custom0]
...
```

In this case we can't use

```nix
org.gnome.settings-daemon.plugins.media-keys.custom-keybindings = [ "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/" ];
media-keys.custom-keybindings.custom0 = { ... };
```

but we can use

```nix
org.gnome.settings-daemon.plugins.media-keys.custom-keybindings = [ "/org/gnome/settings-daemon/plugins/media-keys/custom-keybindings/custom0/" ];
media-keys."custom-keybindings/custom0" = { ... };
```

instead. And if you want you can always use the `"path/to/the/directory"` format.

Reply:
The `subpathLocks` doesn't work with the slash format but this is expected.

Reply:
Well, currently, the option docs do not really mention slash format at all. But if we admit it, and it appears to work just as well as the nested format, as a user, I would expect it to work with locks as well.

But I am starting to think that maybe we should just use the slash format everywhere and use `lib.splitString` in the lock generator. I do not really see much of a benefit of the nested format for users and the slash format more closely resembles the keyfile contents so it will hopefully be less confusing.

Reply:
I can add support for that. Maybe it's easier to use something like `startswith` to generate locks.

Reply:
> The slash format is native to dconf keyfiles but the nested format is native to nix.

Yes, NixOS modules use nested attribute sets for their options but dconf is not NixOS. When modelling a domain, we should stick as closely as possible to it.

> With nested format, it's easier to take advantage of nix features without dealing with key strings. The lockSubpaths is an example.

I think we should leave `lockSubpaths` for a future PR. I do not like modules mixing custom NixOS-specific options with upstream ones and it could be easily replicated with something like:

```
rec {
  settings = {
    "foo/bar/baz" = {
      k1 = true;
    };
    "foo/bar/qux".k2 = true;
    "foo/miz/ter".k3 = "str";
  };
  locks =
    let
      lockedSubpaths = [
        "foo/bar"
      ];
      shouldBeLocked = path: lib.any (lockedSubpath: lib.hasPrefix "${lockedSubpath}/"  "${path}/") lockedSubpaths;
    in
    lib.concatMapAttrs (dirPath: keys: lib.concatMapAttrs (key: val: let path = "${dirPath}/${key}"; in { ${path} = shouldBeLocked path; }) keys) settings;
}
```

(In most cases, `shouldBeLocked` will just be `lib.const true` anyway.)

Reply:
Also reading https://discourse.nixos.org/t/write-key-value-using-lib-hm-gvariant-for-home-manager/31234, I realized we might eventually want `mkValue` to turn nix attrsets into GVariant dictionaries with string keys. If we allow nested keys, it would not be possible.